crates/pop-wallet-integration/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+	pub struct InstantiateArgs {
+		constructor: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+		salt: Option<Vec<u8>>,
+	}
+
+	/// Arguments for calling a contract.
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct CallArgs {
+		address: String,
+		message: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+	}
+
+	#[cfg(test)]
+	mod tests {
+		use super::*;
+		use crate::*;
+		use serde_json::json;
+
+		#[test]
+		fn json_serialize_parachain_data_works() {
+			let data = Data {
+				chain_rpc: "localhost:9944".to_string(),
+				data_type: DataType::Parachain(vec![0, 1, 2, 3]),
+			};
+			let json_data = serde_json::to_value(&data).unwrap();
+
+			assert_eq!(
+				json_data,
+				json!({"chain_rpc": "localhost:9944", "data_type": {"Parachain": [0,1,2,3]}})
+			);
+		}
+		#[test]
+		fn json_serialize_contract_upload_data_works() {
+			let data = Data {
+				chain_rpc: "localhost:9944".to_string(),
+				data_type: DataType::Contract(ContractArgs {
+					call_type: ContractCallType::Upload,
+					storage_deposit_limit: Some("1234".to_string()),
+					code: Some(vec![0, 1, 2]),
+				}),
+			};
+			let json_data = serde_json::to_value(&data).unwrap();
+
+			assert_eq!(
+				json_data,
+				json!({"chain_rpc": "localhost:9944", "data_type": {"Contract": {
+					"call_type": "Upload",
+					"storage_deposit_limit": "1234",
+					"code": [0,1,2]
+				}}})
+			);
+		}
+
+		// TODO: after high level review, complete serialization tests.
+	}
+}
+use crate::contracts::ContractArgs;
+
+struct StateHandler {
+	shutdown_tx: Option<oneshot::Sender<()>>,
+	signed_payload: Option<String>,
+}
+
+/// Manages the wallet integration for secure signing of transactions.
+pub struct WalletIntegrationManager {
+	frontend_path: PathBuf,
+	// cloning can be expensive (e.g. contract code stored in-memory)
+	data: Arc<Data>,
+	signed_payload: Option<String>,
+}
+
+impl WalletIntegrationManager {
+	/// - frontend_path: Path to the wallet-integration frontend.
+	/// - data: Data to be sent to the frontend for signing.
+	pub fn new(frontend_path: PathBuf, data: Data) -> Self {
+		Self { frontend_path, data: Arc::new(data), signed_payload: Default::default() }
+	}
+
+	/// Serves the wallet-integration frontend and an API for the wallet to get
+	/// the necessary data and submit the signed payload.
+	pub async fn run(&mut self) {
+		// used to signal shutdown.
+		let (tx, rx) = oneshot::channel();
+
+		// shared state between routes. Will be used to store the signed payload.
+		let state =
+			Arc::new(Mutex::new(StateHandler { shutdown_tx: Some(tx), signed_payload: None }));
+
+		// will shutdown when the signed payload is received
+		let app = Router::new()
+			// cloning Arcs is cheap
+			.route("/data", get(routes::get_data_handler).with_state(self.data.clone()))
+			.route("/submit", post(routes::handle_submit).with_state(state.clone()))
```

Comment:
What about routes for error reporting, cancellations from the wallet?

Reply:
Thinking out loud. What types of errors might there be in the frontend?
- Wallet cancellation -> problem for the frontend.
- Invalid call data -> theoretically should not happen if pop-cli validates. However, still worthwhile to report to CLI.
- Unable to connect to RPC -> report to CLI
- Dry-run failure

Reply:
https://github.com/r0gue-io/pop-cli/pull/362/commits/54837ae29c45a89e28247c0b0bf30b2efa8c1a3d